### PR TITLE
fix(datastore): data lost when save table with error

### PR DESCRIPTION
**Pull request by goldenxinxing**
## Description

#1916 

1. [Client] add ut for python sdk(dataset and datastore with data of SwObject)
2. [Server] fix lost data bug when save table with error
3. [Server] fix load data exception when data is null at the step of insertRecords 
4. [Server] add retry logic when load is invalid

## Modules
- [ ] UI
- [x] Controller
- [ ] Agent
- [x] Client
- [ ] Python-SDK
- [ ] Others

## Checklist
- [ ] run code format and lint check
- [x] add unit test
- [ ] add necessary doc

### CODE REVIEW

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/parquet/SwWriteSupport.java [hunk 1]
```diff
@@ -76,8 +77,18 @@ public void prepareForWrite(RecordConsumer recordConsumer) {
...
+            this.recordConsumer.endMessage();
+        } catch (Throwable e) {
+            error = true;
+            throw e;
+        }
+    }
+
+    @Override
+    public FinalizedWriteContext finalizeWrite() {
+        return new FinalizedWriteContext(Map.of(SwReadSupport.ERROR_FLAG_KEY, String.valueOf(error)));
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/datastore/parquet/SwWriteSupport.java:92]: How about writing a success flag? We may not be able to catch all errors.

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/parquet/SwReadSupport.java [hunk 2]
```diff
@@ -56,6 +58,10 @@ public RecordMaterializer<Map<String, Object>> prepareForRead(Configuration conf
             Map<String, String> metadata,
             MessageType messageType,
             ReadContext readContext) {
+        var errorFlag = metadata.get(ERROR_FLAG_KEY);
+        if (Boolean.parseBoolean(errorFlag)) {
+            throw new SwProcessException(ErrorType.DATASTORE, "the file is invalid, ignore it!");
+        }
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/datastore/parquet/SwReadSupport.java:64]: Treat the file with `null` flag as the legacy version?

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java [hunk 3]
```diff
@@ -178,8 +181,15 @@ public void save() {
                     writer.write(recordMap);
                 }
             }
-        } catch (IOException e) {
-            throw new SwProcessException(ErrorType.DATASTORE, "failed to save " + this.tableName, e);
+            builder.success();
+        } catch (Throwable e) {
+            builder.error();
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java]: Mark it with an error when init, and no need to change the try segment?

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java [hunk 4]
```diff
@@ -145,7 +146,7 @@ private void load() {
     }
```
Comment (jialeicui) [inline server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java:147]: Try loading the previous file when fail?

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java (hunk 3 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java]: done

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/parquet/SwReadSupport.java (hunk 2 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/datastore/parquet/SwReadSupport.java]: yes

File: server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java (hunk 4 above)
Comment (goldenxinxing) [reply server/controller/src/main/java/ai/starwhale/mlops/datastore/impl/MemoryTableImpl.java:160]: done

**Comment by tianweidut** [comment]
LGTM

**Review by jialeicui — APPROVED** [review]
LGTM

**Pull request merged**